amivapi/auth/link_methods.py

```diff
@@ -0,0 +1,185 @@
+        if link in links:
+            links[link]['methods'] = res_links
+
+
+def add_permitted_methods_after_insert(resource, items):
+    """Add link methods with an on_inserted hook."""
+    if isinstance(resource_auth(resource), AmivTokenAuth):
+        for item in items:
+            add_methods_to_item_links(resource, item)
+
+
+def add_permitted_methods_after_fetch_item(resource, item):
+    """Add link methods with an on_fetched_item hook."""
+    if isinstance(resource_auth(resource), AmivTokenAuth):
+        add_methods_to_item_links(resource, item)
+
+
+def add_permitted_methods_after_fetch_resource(resource, response):
+    """Add link methods with an on_fetched_resource hook."""
+    if isinstance(resource_auth(resource), AmivTokenAuth):
+        # Item links
+        for item in response['_items']:
+            add_methods_to_item_links(resource, item)
+
+        # Resource links
+        add_methods_to_resource_links(resource, response)
+
+
+def _get_data(response):
+    """Get response data as dict.
+
+    Encoding/Decoding necessary for compatibility with both python 2 and 3.
+    """
+    return json.loads(response.get_data().decode('utf-8'))
+
+
+def _set_data(response, data):
+    """Jsonify dict and set as response data."""
+    response.set_data(json.dumps(data))
+
+
+def add_permitted_methods_after_update(resource, request, response):
+    """Add link methods with an on_post_PATCH hook.
+
+    The on_updated hook doesn't work since it doesn't include the links.
+
+    This hook will also be called for errors which do not contain _links.
+    => Make sure to only add methods for successful patches (status 200 only)
+    """
+    if (response.status_code == 200) and \
+            isinstance(resource_auth(resource), AmivTokenAuth):
+        item_data = _get_data(response)
+
+        add_methods_to_item_links(resource, item_data)
+
+        _set_data(response, item_data)
+
+
+def add_permitted_methods_for_home(resource, request, response):
+    """Add link methods to home endpoint with an on_post_GET hook.
+
+    The home endpoint doesn't call any database hooks and no on_pre_GET hook.
+    Therefore authentication needs to be done manually so we can check
+    permissions.
+    """
+    if resource is None:
+        authenticate()
+
+        data = _get_data(response)
+
+        try:
+            links = data['_links']['child']
+        except KeyError:
+            # Other endpoints like `schemaendpoint` might end u here, but don't
+            # have the same 'link' layout as home, so we can just ignore them
+            pass
+        else:
+            # Add links for home
+            for res_link in links:
+                res_name = res_link['title']  # title equals resource
+                if isinstance(resource_auth(res_name), AmivTokenAuth):
+                    check_if_admin(res_name)
+                    res_link['methods'] = _get_resource_methods(res_name)
+
+            _set_data(response, data)
```

Comment:
After going through this file, I wonder if there is any possibility to write this more clearly. All the different methods are really confusing and it is hard to get an overview if at some point somebody wants to change comething. But as long as it works I would simply add this as issue 😉